core/vm/contracts.go

```diff
@@ -360,140 +361,245 @@ type bigModExp struct {
 	eip7883 bool
 }
 
+// Constants for modexp gas calculation
+const (
+	byzantiumMultiplier = 8
+	berlinMultiplier    = 8
+	osakaMultiplier     = 16
+
+	byzantiumDivisor = 20
+	berlinDivisor    = 3
+	osakaDivisor     = 3
+
+	berlinMinGas = 200
+	osakaMinGas  = 500
+)
+
 var (
-	big1      = big.NewInt(1)
-	big3      = big.NewInt(3)
-	big7      = big.NewInt(7)
-	big20     = big.NewInt(20)
-	big32     = big.NewInt(32)
-	big64     = big.NewInt(64)
-	big96     = big.NewInt(96)
-	big480    = big.NewInt(480)
-	big1024   = big.NewInt(1024)
-	big3072   = big.NewInt(3072)
-	big199680 = big.NewInt(199680)
+	// uint256 constants for modexp calculations
+	u256_16     = uint256.NewInt(16)
+	u256_480    = uint256.NewInt(480)
+	u256_199680 = uint256.NewInt(199680)
+
+	// divisors as uint256
+	u256_byzantiumDivisor = uint256.NewInt(byzantiumDivisor)
+	u256_berlinDivisor    = uint256.NewInt(berlinDivisor)
+	u256_osakaDivisor     = uint256.NewInt(osakaDivisor)
 )
 
-// modexpMultComplexity implements bigModexp multComplexity formula, as defined in EIP-198
+// byzantiumMultComplexity implements bigModexp multComplexity formula, as defined in EIP-198
 //
 //	def mult_complexity(x):
 //		if x <= 64: return x ** 2
 //		elif x <= 1024: return x ** 2 // 4 + 96 * x - 3072
 //		else: return x ** 2 // 16 + 480 * x - 199680
 //
 // where is x is max(length_of_MODULUS, length_of_BASE)
-func modexpMultComplexity(x *big.Int) *big.Int {
+func byzantiumMultComplexity(x uint64) *uint256.Int {
 	switch {
-	case x.Cmp(big64) <= 0:
-		x.Mul(x, x) // x ** 2
-	case x.Cmp(big1024) <= 0:
-		// (x ** 2 // 4 ) + ( 96 * x - 3072)
-		x = new(big.Int).Add(
-			new(big.Int).Rsh(new(big.Int).Mul(x, x), 2),
-			new(big.Int).Sub(new(big.Int).Mul(big96, x), big3072),
-		)
+	case x <= 64:
+		return new(uint256.Int).SetUint64(x * x)
+	case x <= 1024:
+		// x^2 / 4 + 96*x - 3072
+		result := x*x/4 + 96*x - 3072
+		return new(uint256.Int).SetUint64(result)
 	default:
-		// (x ** 2 // 16) + (480 * x - 199680)
-		x = new(big.Int).Add(
-			new(big.Int).Rsh(new(big.Int).Mul(x, x), 4),
-			new(big.Int).Sub(new(big.Int).Mul(big480, x), big199680),
-		)
+		// For large x, use uint256 arithmetic to avoid overflow
+		// x^2 / 16 + 480*x - 199680
```

Comment:
x^2 is at most 128 bits, adding 480*x - 199680 still keep us within 256 bits, so we can use a u256 here